### PR TITLE
Add lua pattern matching to core:text with tests

### PR DESCRIPTION
**Pull request by Skytrias**
Implements lua pattern matching (with utf8 support) based on the original implementation at https://github.com/lua/lua/blob/master/lstrlib.c#L573

This implementation uses `int` indices since it sometimes has to step back utf8 codepoints. 

There is also a helper `Matcher` struct which makes life easier to construct quick matches + optional captures - stores and additional info you *might* want

### CODE REVIEW

File: core/text/lua/strlib.odin [hunk 1]
```diff
@@ -0,0 +1,1040 @@
+package strlib
+
+import "core:runtime"
+import "core:unicode"
+import "core:unicode/utf8"
+import "core:strings"
+
+MAXCAPTURES :: 32
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:8]: Try to adhere to the Odin coding style

File: core/text/lua/strlib.odin [hunk 2]
```diff
@@ -0,0 +1,1040 @@
...
+	Rune_Error,
+	Match_Invalid,
+}
+
+L_ESC :: '%'
+CAP_POSITION :: -2
+CAP_UNFINISHED :: -1
+INVALID :: -1
+
+MatchState :: struct {
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:35]: Try to adhere to the Odin naming style

File: core/text/lua/strlib.odin [hunk 3]
```diff
@@ -0,0 +1,1040 @@
...
+islower :: proc(c: rune) -> bool {
+	return unicode.is_lower(c)
+}
+
+isupper :: proc(c: rune) -> bool {
+	return unicode.is_upper(c)
+}
+
+isgraph :: proc(c: rune) -> bool {
+	return unicode.is_digit(c) || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:85]: Add the short circuiting for ascii first.

File: core/text/lua/strlib.odin [hunk 4]
```diff
@@ -0,0 +1,1040 @@
...
+		case 'u': res = isupper(c)
+		case 'w': res = isalnum(c)
+		case 'x': res = isxdigit(c)
+		case: return cl == c
+	}
+
+	return islower(cl) ? res : !res
+}
+
+isalpha :: proc(c: rune) -> bool {
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:60]: Try to adhere to the Odin naming style

File: core/text/lua/strlib.odin [hunk 5]
```diff
@@ -0,0 +1,1040 @@
...
+		case L_ESC: {
+			pchar_next, _ := utf8_peek(ms.pattern[p + psize:]) or_return
+			matched = match_class(schar, pchar_next)
+		}
+		case '[': {
+			matched = matchbracketclass(ms, schar, p, ep - 1) or_return
+		}
+		case: {
+			matched = schar == pchar
+		}
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:282]: Keep to the Odin style by not indenting `case` and not putting the cases in `{}`

File: core/text/lua/strlib.odin [hunk 6]
```diff
@@ -0,0 +1,1040 @@
...
+	s_begin := s
+	cont := 1
+	s := s + ssize
+	begin := pchar
+	end, _ := utf8_peek(ms.pattern[p + psize:]) or_return
+
+	for s < len(ms.src) {
+		ch := utf8_advance(ms.src, &s) or_return
+
+		if ch == end {
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:310]: Make a `switch` statement would be better?

File: core/text/lua/strlib.odin [hunk 7]
```diff
@@ -0,0 +1,1040 @@
...
+		case '(': {
+			if p + 1 < len(ms.pattern) && ms.pattern[p + 1] == ')' {
+				s = start_capture(ms, s, p + 2, CAP_POSITION) or_return
+			} else {
+				s = start_capture(ms, s, p + 1, CAP_UNFINISHED) or_return
+			}
+		}
+
+		case ')': {
+			s = end_capture(ms, s, p + 1) or_return
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:434]: Keep to the Odin style by not indenting case and not putting the cases in {}

File: core/text/lua/strlib.odin [hunk 8]
```diff
@@ -0,0 +1,1040 @@
...
+	single_matched, ssize := singlematch(ms, s, p, ep) or_return
+
+	if !single_matched {
+		epc := ep < len(ms.pattern) ? ms.pattern[ep] : 0
+
+		if epc == '*' || epc == '?' || epc == '-' {
+			return match(ms, s, ep + 1)
+		} else {
+			s = INVALID
+		}
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:532]: Maybe this?

```odin
switch  epc {
case '*', '?', '-':
    return match(ms, s, ep + 1)
case:
    s = INVALID
}

File: core/text/lua/strlib.odin [hunk 9]
```diff
@@ -0,0 +1,1040 @@
...
+	return s, .OK
+}
+
+push_onecapture :: proc(
+	ms: ^MatchState, 
+	i: int, 
+	s: int,
+	e: int,
+	matches: []Match,
+) -> (err: Error) {
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:574]:
```odin
push_onecapture :: proc(ms: ^Match_State, i, s, e: int, matches: []Match) -> (err: Error) {...}
```

**Comment by Skytrias** [comment]
styling issues should be resolved :+1:

File: core/text/lua/strlib.odin [hunk 10]
```diff
@@ -0,0 +1,949 @@
+package strlib
```
Comment (gingerBill) [inline core/text/lua/strlib.odin:1]: `package text_match`
and change the directory to `core:text/match`

File: core/text/match/strlib.odin [hunk 11]
```diff
@@ -0,0 +1,949 @@
+package strlib
```
Comment (gingerBill) [inline core/text/match/strlib.odin:1]: I'd recommend changing the name of this package to `package text_match` to minimize ABI issues.

File: core/text/match/strlib.odin [hunk 12]
```diff
@@ -0,0 +1,949 @@
...
+	for i in 0..<nlevels {
+		push_onecapture(ms, i, s, e, matches) or_return
+	}
+
+	return
+}
+
+// SPECIALS := "^$*+?.([%-"
+// all special characters inside a small ascii array
+SPECIALS_TABLE := [256]u8 { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }
```
Comment (gingerBill) [inline core/text/match/strlib.odin:561]: Any reason this is not just the following?
```odin
SPECIALS_TABLE := [256]bool{
    '^' = true,
    '$' = true,
    '*' = true,
    '+' = true,
    '?' = true,
    '.' = true,
    '(' = true,
    '[' = true,
    '%' = true,
    '-' = true,
}
```

File: core/text/match/strlib.odin [hunk 13]
```diff
@@ -0,0 +1,949 @@
...
+
+// helper call to quick search for special characters
+index_special :: proc(text: string) -> int {
+	// TODO is this utf8 safe?
+	for i in 0..<len(text) {
+		if SPECIALS_TABLE[text[i]] == 1 {
```
Comment (gingerBill) [inline core/text/match/strlib.odin:567]: As with the above being a `bool` now, this can be a simpler check.

**Pull request merged**